File: app/src/main/java/msr/mirudl/app/UpdateChecker.java
package msr.mirudl.app;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Handler;
import android.os.Looper;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.Locale;
import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

/** Checks the app's GitHub repo for a newer tagged release than the one currently installed. */
public final class UpdateChecker {

    private static final String REPO = "msrofficial/MiruDL-App";
    private static final String API_URL = "https://api.github.com/repos/" + REPO + "/releases/latest";

    private static final String PREFS_NAME = "mirudl_update_checker";
    private static final String KEY_TAG = "cached_tag";
    private static final String KEY_BODY = "cached_body";
    private static final String KEY_URL = "cached_url";
    private static final String KEY_APK_URL = "cached_apk_url";
    private static final String KEY_LAST_CHECK = "last_check_at";

    /** How long a cached release check is trusted before re-hitting the network on app start. */
    private static final long CACHE_TTL_MS = 3 * TimeUnit.HOURS.toMillis(1);

    private static final OkHttpClient CLIENT = new OkHttpClient.Builder()
            .connectTimeout(10, TimeUnit.SECONDS)
            .readTimeout(10, TimeUnit.SECONDS)
            .build();

    private UpdateChecker() {}

    public static class ReleaseInfo {
        public String tag;
        public String changelog;
        public String htmlUrl;
        public String apkUrl; // direct .apk asset link, if the release has one
    }

    public interface Callback {
        /** Called on the main thread. {@code info} is null if there's no newer version or the check failed. */
        void onResult(ReleaseInfo info);
    }

    /** Always hits the network. Used by the manual "Check for Updates" action in Settings. */
    public static void checkNow(Context context, Callback callback) {
        Context appCtx = context.getApplicationContext();
        new Thread(() -> {
            ReleaseInfo remote = fetchLatestRelease();
            if (remote != null) cacheRelease(appCtx, remote);
            boolean newer = remote != null && isNewerVersion(remote.tag, currentVersion(appCtx));
            postResult(callback, newer ? remote : null);
        }).start();
    }

    /**
     * Used on app startup. Reuses a recent cached result instead of hitting the network on every
     * launch, but the newer-version verdict is always re-evaluated against the cache — so the
     * reminder keeps appearing on every app open until the user actually updates.
     */
    public static void checkOnStartup(Context context, Callback callback) {
        Context appCtx = context.getApplicationContext();
        SharedPreferences prefs = appCtx.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        long lastCheck = prefs.getLong(KEY_LAST_CHECK, 0);
        boolean stale = System.currentTimeMillis() - lastCheck > CACHE_TTL_MS;

        if (!stale) {
            evaluateCached(appCtx, callback);
            return;
        }

        new Thread(() -> {
            ReleaseInfo remote = fetchLatestRelease();
            if (remote != null) {
                cacheRelease(appCtx, remote);
                boolean newer = isNewerVersion(remote.tag, currentVersion(appCtx));
                postResult(callback, newer ? remote : null);
            } else {
                evaluateCached(appCtx, callback); // network failed — fall back to last known cache
            }
        }).start();
    }

    private static void evaluateCached(Context appCtx, Callback callback) {
        SharedPreferences prefs = appCtx.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        String tag = prefs.getString(KEY_TAG, null);
        if (tag == null || !isNewerVersion(tag, currentVersion(appCtx))) {
            postResult(callback, null);
            return;
        }
        ReleaseInfo info = new ReleaseInfo();
        info.tag = tag;
        info.changelog = prefs.getString(KEY_BODY, "");
        info.htmlUrl = prefs.getString(KEY_URL, null);
        info.apkUrl = prefs.getString(KEY_APK_URL, null);
        postResult(callback, info);
    }

    private static void postResult(Callback callback, ReleaseInfo info) {
        if (callback == null) return;
        new Handler(Looper.getMainLooper()).post(() -> callback.onResult(info));
    }

    private static ReleaseInfo fetchLatestRelease() {
        try {
            Request request = new Request.Builder()
                    .url(API_URL)
                    .header("Accept", "application/vnd.github+json")
                    .header("User-Agent", "MiruDL-App")
                    .build();
            try (Response response = CLIENT.newCall(request).execute()) {
                if (!response.isSuccessful() || response.body() == null) return null;
                JSONObject json = new JSONObject(response.body().string());

                ReleaseInfo info = new ReleaseInfo();
                info.tag = json.optString("tag_name", null);
                info.changelog = json.optString("body", "").trim();
                info.htmlUrl = json.optString("html_url", null);

                JSONArray assets = json.optJSONArray("assets");
                if (assets != null) {
                    for (int i = 0; i < assets.length(); i++) {
                        JSONObject asset = assets.getJSONObject(i);
                        String name = asset.optString("name", "");
                        if (name.toLowerCase(Locale.US).endsWith(".apk")) {
                            info.apkUrl = asset.optString("browser_download_url", null);
                            break;
                        }
                    }
                }
                return (info.tag != null) ? info : null;
            }
        } catch (Exception e) {
            return null;
        }
    }

    private static void cacheRelease(Context appCtx, ReleaseInfo info) {
        appCtx.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE).edit()
                .putString(KEY_TAG, info.tag)
                .putString(KEY_BODY, info.changelog)
                .putString(KEY_URL, info.htmlUrl)
                .putString(KEY_APK_URL, info.apkUrl)
                .putLong(KEY_LAST_CHECK, System.currentTimeMillis())
                .apply();
    }

    private static String currentVersion(Context context) {
        try {
            return context.getPackageManager().getPackageInfo(context.getPackageName(), 0).versionName;
        } catch (Exception e) {
            return "0";
        }
    }

    /** True if remoteTag represents a newer version than localVersion (e.g. "v1.0.110" > "1.0.108"). */
    public static boolean isNewerVersion(String remoteTag, String localVersion) {
        int[] r = normalize(remoteTag);
        int[] l = normalize(localVersion);
        int len = Math.max(r.length, l.length);
        for (int i = 0; i < len; i++) {
            int rv = i < r.length ? r[i] : 0;
            int lv = i < l.length ? l[i] : 0;
            if (rv != lv) return rv > lv;
        }
        return false;
    }

    private static int[] normalize(String v) {
        if (v == null) return new int[]{0};
        v = v.trim();
        if (v.toLowerCase(Locale.US).startsWith("v")) v = v.substring(1);
        String[] parts = v.split("\\.");
        int[] out = new int[parts.length];
        for (int i = 0; i < parts.length; i++) {
            try {
                out[i] = Integer.parseInt(parts[i].replaceAll("[^0-9]", ""));
            } catch (Exception e) {
                out[i] = 0;
            }
        }
        return out;
    }
}

File: shared/src/commonMain/kotlin/msr/mirudl/shared/network/UpdateChecker.kt
package msr.mirudl.shared.network

import io.ktor.client.request.*
import io.ktor.client.statement.bodyAsText
import kotlinx.serialization.json.*
import msr.mirudl.shared.storage.AppStorage

/**
 * Shared port of `msr.mirudl.app.UpdateChecker` (Java).
 *
 * Platform-specific concerns (HTTP engine, background thread dispatch,
 * `Handler`/`Looper` posting) live in `UpdateCheckerAndroid` — this
 * file contains only pure logic and cache helpers.
 */
object UpdateChecker {

    private const val REPO = "msrofficial/MiruDL-App"
    private const val API_URL = "https://api.github.com/repos/$REPO/releases/latest"

    private const val KEY_TAG = "cached_tag"
    private const val KEY_BODY = "cached_body"
    private const val KEY_URL = "cached_url"
    private const val KEY_APK_URL = "cached_apk_url"
    private const val KEY_LAST_CHECK = "last_check_at"

    const val CACHE_TTL_MS: Long = 3 * 60 * 60 * 1000 // 3 hours

    class ReleaseInfo @JvmOverloads constructor(
        @JvmField var tag: String = "",
        @JvmField var changelog: String = "",
        @JvmField var htmlUrl: String? = null,
        @JvmField var apkUrl: String? = null
    )

    suspend fun fetchLatestRelease(): ReleaseInfo? {
        return try {
            val response = HttpClientProvider.get().get(API_URL) {
                header("Accept", "application/vnd.github+json")
                header("User-Agent", "MiruDL-App")
            }
            if (response.status.value != 200) return null
            val body = response.bodyAsText()
            val root = Json.parseToJsonElement(body) as? JsonObject ?: return null
            val tag = (root["tag_name"] as? JsonPrimitive)?.content?.trim()
                ?: return null
            val changelog = (root["body"] as? JsonPrimitive)?.content?.trim() ?: ""
            val htmlUrl = (root["html_url"] as? JsonPrimitive)?.content?.trim()

            var apkUrl: String? = null
            val assets = root["assets"] as? JsonArray
            if (assets != null) {
                for (element in assets) {
                    val obj = element as? JsonObject ?: continue
                    val name = (obj["name"] as? JsonPrimitive)?.content ?: ""
                    if (name.lowercase().endsWith(".apk")) {
                        apkUrl = (obj["browser_download_url"] as? JsonPrimitive)?.content
                        break
                    }
                }
            }

            ReleaseInfo(
                tag = tag,
                changelog = changelog,
                htmlUrl = htmlUrl,
                apkUrl = apkUrl
            )
        } catch (e: Exception) {
            null
        }
    }

    fun isNewerVersion(remoteTag: String, localVersion: String): Boolean {
        val r = normalize(remoteTag)
        val l = normalize(localVersion)
        val len = maxOf(r.size, l.size)
        for (i in 0 until len) {
            val rv = if (i < r.size) r[i] else 0
            val lv = if (i < l.size) l[i] else 0
            if (rv != lv) return rv > lv
        }
        return false
    }

    fun normalize(v: String?): IntArray {
        if (v == null) return intArrayOf(0)
        var s = v.trim()
        if (s.lowercase().startsWith("v")) s = s.substring(1)
        val parts = s.split(".")
        return IntArray(parts.size) { i ->
            parts[i].replace(Regex("[^0-9]"), "").toIntOrNull() ?: 0
        }
    }

    fun readLastCheck(storage: AppStorage): Long {
        return storage.getLong(KEY_LAST_CHECK, 0L)
    }

    fun readCache(storage: AppStorage): ReleaseInfo? {
        val tag = storage.getString(KEY_TAG, "")
        if (tag.isEmpty() || !isNewerVersion(tag, "")) return null
        return ReleaseInfo(
            tag = tag,
            changelog = storage.getString(KEY_BODY, ""),
            htmlUrl = storage.getString(KEY_URL, "").ifEmpty { null },
            apkUrl = storage.getString(KEY_APK_URL, "").ifEmpty { null }
        )
    }

    fun writeCache(storage: AppStorage, info: ReleaseInfo) {
        storage.setString(KEY_TAG, info.tag)
        storage.setString(KEY_BODY, info.changelog)
        storage.setString(KEY_URL, info.htmlUrl ?: "")
        storage.setString(KEY_APK_URL, info.apkUrl ?: "")
        storage.setLong(KEY_LAST_CHECK, System.currentTimeMillis())
    }
}
